File: src/main/java/com/openlattice/chronicle/services/enrollment/EnrollmentManager.java
package com.openlattice.chronicle.services.enrollment;

import com.google.common.base.Optional;
import com.openlattice.chronicle.data.ParticipationStatus;
import com.openlattice.chronicle.sources.Datasource;
import org.apache.olingo.commons.api.edm.FullQualifiedName;

import java.util.Map;
import java.util.Set;
import java.util.UUID;

/**
 * @author alfoncenzioka &lt;alfonce@openlattice.com&gt;
 */
public interface EnrollmentManager {

    UUID registerDatasource(
            UUID organizationId,
            UUID studyId,
            String participantId,
            String datasourceId,
            Optional<Datasource> datasource );

    UUID getDeviceEntityKeyId( UUID organizationId, UUID studyId, String participantId, String datasourceId );

    boolean isKnownDatasource( UUID organizationId, UUID studyId, String participantId, String datasourceId );

    boolean isKnownParticipant( UUID organizationId, UUID studyId, String participantId );

    Map<FullQualifiedName, Set<Object>> getParticipantEntity(
            UUID organizationId,
            UUID studyId,
            UUID participantEntityId );

    ParticipationStatus getParticipationStatus( UUID organizationId, UUID studyId, String participantId );

    boolean isNotificationsEnabled( UUID organizationId, UUID studyId );

    UUID getParticipantEntityKeyId( UUID organizationId, UUID studyId, String participantId );

    UUID getStudyEntityKeyId( UUID organizationId, UUID studyId );
}

File: src/main/kotlin/com/openlattice/chronicle/services/enrollment/EnrollmentManager.kt
package com.openlattice.chronicle.services.enrollment

import com.openlattice.chronicle.data.ParticipationStatus
import com.openlattice.chronicle.participants.Participant
import com.openlattice.chronicle.sources.SourceDevice
import com.openlattice.chronicle.sources.SourceDeviceType
import java.sql.Connection
import java.util.*

/**
 *
 * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
 */
/**
 * @author alfoncenzioka &lt;alfonce@openlattice.com&gt;
 */
interface EnrollmentManager {
    fun registerDevice(
        studyId: UUID,
        participantId: String,
        sourceDeviceId: String,
        sourceDevice: SourceDevice
    ): UUID

    fun isKnownDatasource(studyId: UUID, participantId: String, sourceDeviceId: String): Boolean
    fun isKnownParticipant(studyId: UUID, participantId: String): Boolean

    fun getParticipant(studyId: UUID, participantId: String): Participant
    fun getParticipationStatus(studyId: UUID, participantId: String): ParticipationStatus
    fun getStudyParticipantIds(studyId: UUID): Set<String>
    fun getStudyParticipants(studyId: UUID): Set<Participant>
    fun studyExists(studyId: UUID): Boolean
    fun getOrganizationIdForStudy(studyId: UUID): UUID
    fun getDeviceId(studyId: UUID, participantId: String, sourceDeviceId: String): UUID

    fun registerParticipant(
        connection: Connection,
        studyId: UUID,
        participantId: String,
        candidateId: UUID,
        participationStatus: ParticipationStatus
    )
}
